cmd/npmExecuteScripts.go

```diff
@@ -0,0 +1,195 @@
+	fileExists(path string) (bool, error)
+	glob(pattern string) (matches []string, err error)
+	getwd() (dir string, err error)
+	chdir(dir string) error
+	getExecRunner() execRunner
+}
+
+type npmExecuteScriptsUtilsBundle struct {
+	projectStructure FileUtils.ProjectStructure
+	fileUtils        FileUtils.Files
+	execRunner       *command.Command
+}
+
+func (u *npmExecuteScriptsUtilsBundle) fileExists(path string) (bool, error) {
+	return u.fileUtils.FileExists(path)
+}
+
+func (u *npmExecuteScriptsUtilsBundle) glob(pattern string) (matches []string, err error) {
+	return doublestar.Glob(pattern)
+}
+
+func (u *npmExecuteScriptsUtilsBundle) getwd() (dir string, err error) {
+	return os.Getwd()
+}
+
+func (u *npmExecuteScriptsUtilsBundle) chdir(dir string) error {
+	return os.Chdir(dir)
+}
+
+func (u *npmExecuteScriptsUtilsBundle) getExecRunner() execRunner {
+	if u.execRunner == nil {
+		u.execRunner = &command.Command{}
+		u.execRunner.Stdout(log.Entry().Writer())
+		u.execRunner.Stderr(log.Entry().Writer())
+	}
+	return u.execRunner
+}
+
+func npmExecuteScripts(config npmExecuteScriptsOptions, telemetryData *telemetry.CustomData) {
+	utils := npmExecuteScriptsUtilsBundle{}
+
+	err := runNpmExecuteScripts(&utils, &config)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
+	}
+}
+func runNpmExecuteScripts(utils npmExecuteScriptsUtilsInterface, options *npmExecuteScriptsOptions) error {
+	execRunner := utils.getExecRunner()
+	packageJSONFiles, err := findPackageJSONFiles(utils)
+	if err != nil {
+		return err
+	}
+
+	oldWorkingDirectory, err := utils.getwd()
+
+	for _, file := range packageJSONFiles {
+		dir := path.Dir(file)
+		err = utils.chdir(dir)
+		if err != nil {
+			return err
+		}
+
+		// set in each directory to respect existing config in rc files
+		err = setNpmRegistries(options, execRunner)
+		if err != nil {
+			return err
+		}
+
+		packageLockExists, yarnLockExists, err := checkIfLockFilesExist(utils)
+		if err != nil {
+			return err
+		}
+		if options.Install {
+			err = installDependencies(dir, packageLockExists, yarnLockExists, execRunner)
+			if err != nil {
+				return err
+			}
+		}
+
+		for _, v := range options.RunScripts {
+			log.Entry().WithField("WorkingDirectory", dir).Info("run-script " + v)
+			err = execRunner.RunExecutable("npm", "run-script", v, "--if-present")
+			if err != nil {
+				return err
+			}
+		}
+		err = utils.chdir(oldWorkingDirectory)
+		if err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+func setNpmRegistries(options *npmExecuteScriptsOptions, execRunner execRunner) error {
+	var environment []string
+	configurableRegistries := []string{"registry", "@sap:registry"}
+	for _, registry := range configurableRegistries {
+		var buffer bytes.Buffer
+		execRunner.Stdout(&buffer)
+		err := execRunner.RunExecutable("npm", "config", "get", registry)
+		if err != nil {
+			return err
+		}
+		preConfiguredRegistry := buffer.String()
+		execRunner.Stdout(log.Entry().Writer())
```

Comment:
I think you should reset `exerRunner.Stdout()`via `defer`, so that it is reset also in the error case. While it doesn't really change anything, since you exit in case of error in the calling code, as far as the function is concerned, it should leave the runner in a valid state, so that calling code /could/ be changed.